File: assistant-voice.js
/**
 * 🎤 VOICE-ENABLED LOCAL ASSISTANT
 * 
 * Press SPACE to start recording, release to stop.
 * Uses OpenAI Whisper API for speech-to-text.
 * 
 * SETUP:
 *   1. Install SoX: https://sourceforge.net/projects/sox/
 *   2. Add SoX to PATH
 *   3. Set OPENAI_API_KEY in .env file
 *   4. npm run voice
 */

import { spawn, exec } from 'child_process';
import * as readline from 'readline';
import chalk from 'chalk';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import OpenAI from 'openai';
import 'dotenv/config';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ═══════════════════════════════════════════════════════════════════════════
// 🔧 Configuration
// ═══════════════════════════════════════════════════════════════════════════

const CONFIG = {
  // Recording settings
  sampleRate: 16000,
  channels: 1,
  audioFile: path.join(__dirname, 'temp_recording.wav'),
  
  // Whisper settings
  whisperModel: 'whisper-1',
  
  // Voice activation
  useVoiceActivation: false, // Set to true for always-listening mode
  silenceThreshold: 0.01,
  silenceDuration: 1500, // ms of silence to stop recording
};

// ═══════════════════════════════════════════════════════════════════════════
// 🎨 Terminal UI
// ═══════════════════════════════════════════════════════════════════════════

const UI = {
  banner() {
    console.log(chalk.cyan(`
╔═══════════════════════════════════════════════════════════════════════╗
║                                                                       ║
║   🎤 VOICE-ENABLED LOCAL ASSISTANT                                    ║
║   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ║
║   Press ENTER to start recording • Press ENTER again to stop         ║
║   Type commands directly • Say "exit" or type "exit" to quit         ║
║                                                                       ║
╚═══════════════════════════════════════════════════════════════════════╝
    `));
  },

  recording() {
    console.log(chalk.red.bold('\n  🔴 RECORDING... (Press ENTER to stop)\n'));
  },

  processing() {
    process.stdout.write(chalk.yellow('  🔄 Transcribing...'));
  },

  transcribed(text) {
    console.log(chalk.magenta(`\n  🎤 You said: "${text}"`));
  },

  thinking() {
    process.stdout.write(chalk.yellow('  ⚡ Executing...'));
  },

  success(msg) {
    console.log(chalk.green(`\n  ✅ ${msg}`));
  },

  error(msg) {
    console.log(chalk.red(`\n  ❌ ${msg}`));
  },

  info(msg) {
    console.log(chalk.cyan(`  ℹ️  ${msg}`));
  },

  response(msg) {
    console.log(chalk.white(`\n  🤖 ${msg}\n`));
  },

  prompt() {
    process.stdout.write(chalk.green('\n  [ENTER=🎤 Voice | Type command]: '));
  }
};

// ═══════════════════════════════════════════════════════════════════════════
// 🎤 Voice Recorder using SoX
// ═══════════════════════════════════════════════════════════════════════════

class VoiceRecorder {
  constructor() {
    this.recording = false;
    this.process = null;
  }

  async checkSoX() {
    return new Promise((resolve) => {
      exec('sox --version', (error) => {
        resolve(!error);
      });
    });
  }

  start() {
    return new Promise((resolve, reject) => {
      this.recording = true;
      
      // Use SoX to record audio
      // Windows: rec command from SoX
      this.process = spawn('rec', [
        '-q',                    // Quiet
        '-r', CONFIG.sampleRate.toString(), // Sample rate
        '-c', CONFIG.channels.toString(),   // Channels
        '-b', '16',              // Bits
        CONFIG.audioFile,        // Output file
      ], {
        stdio: ['pipe', 'pipe', 'pipe']
      });

      this.process.on('error', (err) => {
        this.recording = false;
        reject(new Error(`Failed to start recording. Is SoX installed? Error: ${err.message}`));
      });

      resolve();
    });
  }

  stop() {
    return new Promise((resolve) => {
      if (this.process) {
        this.recording = false;
        
        // Send quit signal to SoX
        this.process.kill('SIGTERM');
        
        // Wait a bit for file to be written
        setTimeout(() => {
          resolve(CONFIG.audioFile);
        }, 500);
      } else {
        resolve(null);
      }
    });
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// 🔊 Speech-to-Text using OpenAI Whisper
// ═══════════════════════════════════════════════════════════════════════════

class WhisperTranscriber {
  constructor() {
    if (!process.env.OPENAI_API_KEY) {
      throw new Error('OPENAI_API_KEY not found in environment. Create a .env file with your key.');
    }
    this.openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY
    });
  }

  async transcribe(audioFile) {
    try {
      const response = await this.openai.audio.transcriptions.create({
        file: fs.createReadStream(audioFile),
        model: CONFIG.whisperModel,
        language: 'en',
      });
      
      // Clean up temp file
      try { fs.unlinkSync(audioFile); } catch {}
      
      return response.text;
    } catch (error) {
      throw new Error(`Transcription failed: ${error.message}`);
    }
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// 🛠️ PowerShellTool (Same as before)
// ═══════════════════════════════════════════════════════════════════════════

class PowerShellTool {
  async execute(command) {
    return new Promise((resolve, reject) => {
      exec(`powershell -Command "${command}"`, (error, stdout, stderr) => {
        if (error) {
          reject(stderr || error.message);
        } else {
          resolve(stdout.trim());
        }
      });
    });
  }

  async openApp(appName) {
    const appMap = {
      'outlook': 'Start-Process outlook',
      'word': 'Start-Process winword',
      'excel': 'Start-Process excel',
      'powerpoint': 'Start-Process powerpnt',
      'teams': 'Start-Process ms-teams:',
      'chrome': 'Start-Process chrome',
      'firefox': 'Start-Process firefox',
      'edge': 'Start-Process msedge',
      'notepad': 'Start-Process notepad',
      'calculator': 'Start-Process calc',
      'explorer': 'Start-Process explorer',
      'terminal': 'Start-Process wt',
      'settings': 'Start-Process ms-settings:',
      'vscode': 'Start-Process code',
      'vs code': 'Start-Process code',
      'visual studio code': 'Start-Process code',
      'spotify': 'Start-Process spotify',
      'discord': 'Start-Process discord',
      'slack': 'Start-Process slack',
    };

    const key = appName.toLowerCase().trim();
    const command = appMap[key];

    if (command) {
      await this.execute(command);
      return `Opening ${appName}...`;
    } else {
      try {
        await this.execute(`Start-Process "${appName}"`);
        return `Trying to open ${appName}...`;
      } catch {
        throw new Error(`Unknown app: ${appName}`);
      }
    }
  }

  async searchWeb(query, engine = 'google') {
    const engines = {
      'google': `https://www.google.com/search?q=`,
      'youtube': `https://www.youtube.com/results?search_query=`,
      'github': `https://github.com/search?q=`,
    };
    const baseUrl = engines[engine] || engines['google'];
    const url = `${baseUrl}${encodeURIComponent(query)}`;
    await this.execute(`Start-Process "${url}"`);
    return `Searching for "${query}"...`;
  }

  async openUrl(url) {
    if (!url.startsWith('http')) url = 'https://' + url;
    await this.execute(`Start-Process "${url}"`);
    return `Opening ${url}...`;
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// 🧠 Command Parser
// ═══════════════════════════════════════════════════════════════════════════

class CommandParser {
  parse(input) {
    const text = input.toLowerCase().trim();
    
    if (['exit', 'quit', 'bye', 'goodbye', 'stop', 'close assistant'].includes(text)) {
      return { action: 'exit' };
    }

    if (['help', 'what can you do', 'commands'].includes(text)) {
      return { action: 'help' };
    }

    // Open app
    const openMatch = text.match(/^(open|launch|start)\s+(.+)$/i);
    if (openMatch) {
      const target = openMatch[2];
      if (target.includes('.com') || target.includes('.org') || target.startsWith('http')) {
        return { action: 'url', target };
      }
      return { action: 'open', app: target };
    }

    // Search
    const searchMatch = text.match(/^(search|google|youtube|github)\s+(.+)$/i);
    if (searchMatch) {
      const engine = searchMatch[1].toLowerCase() === 'search' ? 'google' : searchMatch[1].toLowerCase();
      return { action: 'search', query: searchMatch[2], engine };
    }

    // Go to URL
    const gotoMatch = text.match(/^(go to|goto|navigate to|visit)\s+(.+)$/i);
    if (gotoMatch) {
      return { action: 'url', target: gotoMatch[2] };
    }

    // Time
    if (text.match(/^(what('?s| is) the )?(time|date)/i) || text === 'time') {
      return { action: 'time' };
    }

    // Lock
    if (text.match(/^lock/i)) {
      return { action: 'lock' };
    }

    // PowerShell command
    const psMatch = text.match(/^(run|execute|ps:?)\s+(.+)$/i);
    if (psMatch) {
      return { action: 'command', command: psMatch[2] };
    }

    return { action: 'unknown', input: text };
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// 🤖 Voice Assistant
// ═══════════════════════════════════════════════════════════════════════════

class VoiceAssistant {
  constructor() {
    this.ps = new PowerShellTool();
    this.parser = new CommandParser();
    this.recorder = new VoiceRecorder();
    this.transcriber = null;
    this.running = true;
    this.isRecording = false;
  }

  async init() {
    // Check SoX
    const hasSoX = await this.recorder.checkSoX();
    if (!hasSoX) {
      console.log(chalk.yellow(`
  ⚠️  SoX not found! Voice recording won't work.
  
  To install SoX on Windows:
    1. Download from: https://sourceforge.net/projects/sox/
    2. Install and add to PATH
    3. Restart terminal
  
  You can still type commands manually.
      `));
    }

    // Check OpenAI API key
    try {
      this.transcriber = new WhisperTranscriber();
      UI.info('Whisper API connected ✓');
    } catch (err) {
      console.log(chalk.yellow(`
  ⚠️  ${err.message}
  
  To enable voice:
    1. Create .env file in this folder
    2. Add: OPENAI_API_KEY=sk-your-key-here
    3. Restart the assistant
  
  You can still type commands manually.
      `));
    }
  }

  showHelp() {
    console.log(chalk.cyan(`
  ╔═══════════════════════════════════════════════════════════════════╗
  ║  🎤 VOICE COMMANDS (or type them!)                                ║
  ╠═══════════════════════════════════════════════════════════════════╣
  ║  "Open Outlook"        - Opens Outlook                            ║
  ║  "Open Chrome"         - Opens Chrome                             ║
  ║  "Open VS Code"        - Opens Visual Studio Code                 ║
  ║  "Search [query]"      - Google search                            ║
  ║  "YouTube [query]"     - YouTube search                           ║
  ║  "Go to [website]"     - Open a website                           ║
  ║  "What time is it"     - Show current time                        ║
  ║  "Lock"                - Lock screen                              ║
  ║  "Exit" / "Quit"       - Close assistant                          ║
  ╚═══════════════════════════════════════════════════════════════════╝
    `));
  }

  async execute(parsed) {
    try {
      switch (parsed.action) {
        case 'exit':
          this.running = false;
          UI.response('Goodbye! 👋');
          return;

        case 'help':
          this.showHelp();
          return;

        case 'open':
          UI.thinking();
          const openResult = await this.ps.openApp(parsed.app);
          UI.success(openResult);
          return;

        case 'url':
          UI.thinking();
          const urlResult = await this.ps.openUrl(parsed.target);
          UI.success(urlResult);
          return;

        case 'search':
          UI.thinking();
          const searchResult = await this.ps.searchWeb(parsed.query, parsed.engine);
          UI.success(searchResult);
          return;

        case 'time':
          const now = new Date();
          UI.response(`It's ${now.toLocaleTimeString()} on ${now.toLocaleDateString()}`);
          return;

        case 'lock':
          UI.thinking();
          await this.ps.execute('rundll32.exe user32.dll,LockWorkStation');
          UI.success('Locking screen...');
          return;

        case 'command':
          UI.thinking();
          const result = await this.ps.execute(parsed.command);
          console.log(chalk.cyan('\n  📤 Output:\n'));
          console.log(chalk.white('  ' + (result || '(no output)').split('\n').join('\n  ')));
          return;

        case 'unknown':
          UI.response(`I don't understand "${parsed.input}". Say "help" for commands.`);
          return;
      }
    } catch (err) {
      UI.error(err.message || err);
    }
  }

  async handleVoiceInput() {
    if (!this.transcriber) {
      UI.error('Voice not available. Type your command instead.');
      return null;
    }

    try {
      UI.recording();
      await this.recorder.start();
      
      // Wait for user to press Enter again
      await new Promise(resolve => {
        const onKeypress = () => {
          process.stdin.removeListener('data', onKeypress);
          resolve();
        };
        process.stdin.once('data', onKeypress);
      });

      const audioFile = await this.recorder.stop();
      
      if (!audioFile || !fs.existsSync(audioFile)) {
        UI.error('No audio recorded');
        return null;
      }

      UI.processing();
      const text = await this.transcriber.transcribe(audioFile);
      UI.transcribed(text);
      
      return text;
    } catch (err) {
      UI.error(`Voice error: ${err.message}`);
      return null;
    }
  }

  async run() {
    UI.banner();
    await this.init();

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    // Enable raw mode for key detection
    if (process.stdin.isTTY) {
      process.stdin.setRawMode(false);
    }

    const prompt = () => {
      if (!this.running) {
        rl.close();
        process.exit(0);
        return;
      }

      rl.question(chalk.green('\n  [ENTER=🎤 Voice | Type command]: '), async (input) => {
        input = input.trim();

        // Empty input = voice mode
        if (input === '' && this.transcriber) {
          const voiceText = await this.handleVoiceInput();
          if (voiceText) {
            const parsed = this.parser.parse(voiceText);
            await this.execute(parsed);
          }
        } else if (input) {
          // Text input
          const parsed = this.parser.parse(input);
          await this.execute(parsed);
        }

        if (this.running) {
          prompt();
        } else {
          rl.close();
          process.exit(0);
        }
      });
    };

    prompt();
  }
}

// ═══════════════════════════════════════════════════════════════════════════
// 🚀 START
// ═══════════════════════════════════════════════════════════════════════════

console.clear();
const assistant = new VoiceAssistant();
assistant.run().catch(console.error);

File: assistant_voice.py
"""
🎤 VOICE-ENABLED LOCAL ASSISTANT (No External Tools Required!)

Uses Windows native speech recognition - NO SoX or external installs needed!

SETUP (one-time):
    pip install SpeechRecognition pyaudio

USAGE:
    python assistant_voice.py

Press ENTER to start voice recording, speak, then it auto-detects when you stop.
"""

import subprocess
import os
import sys
from datetime import datetime

# Check and install dependencies
def check_dependencies():
    missing = []
    try:
        import speech_recognition
    except ImportError:
        missing.append('SpeechRecognition')
    
    try:
        import pyaudio
    except ImportError:
        missing.append('pyaudio')
    
    if missing:
        print(f"\n⚠️  Missing packages: {', '.join(missing)}")
        print("\nInstalling automatically...")
        for pkg in missing:
            subprocess.check_call([sys.executable, '-m', 'pip', 'install', pkg, '-q'])
        print("✅ Dependencies installed! Restarting...\n")
        os.execv(sys.executable, [sys.executable] + sys.argv)

check_dependencies()

import speech_recognition as sr

# ═══════════════════════════════════════════════════════════════════════════
# 🎨 Terminal UI
# ═══════════════════════════════════════════════════════════════════════════

class Colors:
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    MAGENTA = '\033[95m'
    WHITE = '\033[97m'
    BOLD = '\033[1m'
    END = '\033[0m'

def banner():
    print(f"""{Colors.CYAN}
╔═══════════════════════════════════════════════════════════════════════╗
║                                                                       ║
║   🎤 VOICE-ENABLED LOCAL ASSISTANT                                    ║
║   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ║
║   Windows Native Speech Recognition • No External Tools Required!    ║
║                                                                       ║
║   Press ENTER to speak • Type commands directly • Say "exit" to quit ║
║                                                                       ║
╚═══════════════════════════════════════════════════════════════════════╝
    {Colors.END}""")

def recording():
    print(f"{Colors.RED}{Colors.BOLD}\n  🔴 LISTENING... (speak now, pause to stop){Colors.END}\n")

def success(msg):
    print(f"{Colors.GREEN}\n  ✅ {msg}{Colors.END}")

def error(msg):
    print(f"{Colors.RED}\n  ❌ {msg}{Colors.END}")

def info(msg):
    print(f"{Colors.CYAN}  ℹ️  {msg}{Colors.END}")

def transcribed(text):
    print(f"{Colors.MAGENTA}\n  🎤 You said: \"{text}\"{Colors.END}")

def response(msg):
    print(f"{Colors.WHITE}\n  🤖 {msg}\n{Colors.END}")

# ═══════════════════════════════════════════════════════════════════════════
# 🎤 Windows Native Speech Recognition
# ═══════════════════════════════════════════════════════════════════════════

class WindowsSpeechRecognizer:
    """Uses Windows built-in speech recognition - NO external tools needed!"""
    
    def __init__(self):
        self.recognizer = sr.Recognizer()
        self.microphone = sr.Microphone()
        
        # Adjust for ambient noise on startup
        print("  🎤 Calibrating microphone...")
        with self.microphone as source:
            self.recognizer.adjust_for_ambient_noise(source, duration=1)
        print("  ✅ Microphone ready!\n")
    
    def listen(self):
        """Listen and transcribe using Windows Speech Recognition (offline!)"""
        with self.microphone as source:
            recording()
            try:
                # Listen with automatic silence detection
                audio = self.recognizer.listen(source, timeout=10, phrase_time_limit=15)
                
                print(f"{Colors.YELLOW}  🔄 Transcribing (Windows Speech Recognition)...{Colors.END}")
                
                # Use Windows native speech recognition (OFFLINE - no API!)
                try:
                    # First try Windows native (offline, fast)
                    text = self.recognizer.recognize_sphinx(audio)
                    return text
                except sr.UnknownValueError:
                    pass
                except Exception:
                    pass
                
                # Fallback to Google (free, online)
                try:
                    text = self.recognizer.recognize_google(audio)
                    return text
                except sr.UnknownValueError:
                    error("Could not understand audio")
                    return None
                except sr.RequestError as e:
                    error(f"Speech service error: {e}")
                    return None
                    
            except sr.WaitTimeoutError:
                error("No speech detected (timeout)")
                return None

# ═══════════════════════════════════════════════════════════════════════════
# 🛠️ PowerShell Command Executor
# ═══════════════════════════════════════════════════════════════════════════

class PowerShellTool:
    def execute(self, command):
        try:
            result = subprocess.run(
                ['powershell', '-Command', command],
                capture_output=True, text=True, timeout=30
            )
            return result.stdout.strip() if result.returncode == 0 else result.stderr.strip()
        except Exception as e:
            return str(e)

    def open_app(self, app_name):
        app_map = {
            'outlook': 'Start-Process outlook',
            'word': 'Start-Process winword',
            'excel': 'Start-Process excel',
            'powerpoint': 'Start-Process powerpnt',
            'teams': 'Start-Process ms-teams:',
            'chrome': 'Start-Process chrome',
            'firefox': 'Start-Process firefox',
            'edge': 'Start-Process msedge',
            'notepad': 'Start-Process notepad',
            'calculator': 'Start-Process calc',
            'explorer': 'Start-Process explorer',
            'terminal': 'Start-Process wt',
            'settings': 'Start-Process ms-settings:',
            'vscode': 'Start-Process code',
            'vs code': 'Start-Process code',
            'visual studio code': 'Start-Process code',
            'spotify': 'Start-Process spotify',
            'discord': 'Start-Process discord',
            'slack': 'Start-Process slack',
            'zoom': 'Start-Process zoom',
            'jira': 'Start-Process "https://rb-tracker.bosch.com"',
        }
        
        key = app_name.lower().strip()
        command = app_map.get(key)
        
        if command:
            self.execute(command)
            return f"Opening {app_name}..."
        else:
            # Try direct
            self.execute(f'Start-Process "{app_name}"')
            return f"Trying to open {app_name}..."

    def search_web(self, query, engine='google'):
        engines = {
            'google': 'https://www.google.com/search?q=',
            'youtube': 'https://www.youtube.com/results?search_query=',
            'github': 'https://github.com/search?q=',
            'bing': 'https://www.bing.com/search?q=',
        }
        base_url = engines.get(engine, engines['google'])
        url = f'{base_url}{query.replace(" ", "+")}'
        self.execute(f'Start-Process "{url}"')
        return f'Searching for "{query}" on {engine}...'

    def open_url(self, url):
        if not url.startswith('http'):
            url = 'https://' + url
        self.execute(f'Start-Process "{url}"')
        return f'Opening {url}...'

# ═══════════════════════════════════════════════════════════════════════════
# 🧠 Command Parser
# ═══════════════════════════════════════════════════════════════════════════

def parse_command(text):
    text = text.lower().strip()
    
    # Exit
    if text in ['exit', 'quit', 'bye', 'goodbye', 'stop', 'close']:
        return {'action': 'exit'}
    
    # Help
    if text in ['help', 'commands', 'what can you do']:
        return {'action': 'help'}
    
    # Open app: "open outlook", "launch chrome"
    import re
    open_match = re.match(r'^(open|launch|start)\s+(.+)$', text, re.IGNORECASE)
    if open_match:
        target = open_match.group(2)
        if any(x in target for x in ['.com', '.org', '.io', 'http']):
            return {'action': 'url', 'target': target}
        return {'action': 'open', 'app': target}
    
    # Search: "search how to code", "youtube funny cats"
    search_match = re.match(r'^(search|google|youtube|github|bing)\s+(.+)$', text, re.IGNORECASE)
    if search_match:
        engine = 'google' if search_match.group(1).lower() == 'search' else search_match.group(1).lower()
        return {'action': 'search', 'query': search_match.group(2), 'engine': engine}
    
    # Go to URL
    goto_match = re.match(r'^(go to|goto|visit|navigate to)\s+(.+)$', text, re.IGNORECASE)
    if goto_match:
        return {'action': 'url', 'target': goto_match.group(2)}
    
    # Time
    if re.match(r'^(what.*time|time|date|what.*date)', text, re.IGNORECASE) or text in ['time', 'date']:
        return {'action': 'time'}
    
    # Lock
    if 'lock' in text:
        return {'action': 'lock'}
    
    # PowerShell command
    ps_match = re.match(r'^(run|execute|ps:?)\s+(.+)$', text, re.IGNORECASE)
    if ps_match:
        return {'action': 'command', 'command': ps_match.group(2)}
    
    return {'action': 'unknown', 'input': text}

# ═══════════════════════════════════════════════════════════════════════════
# 🤖 Main Assistant
# ═══════════════════════════════════════════════════════════════════════════

def show_help():
    print(f"""{Colors.CYAN}
  ╔═══════════════════════════════════════════════════════════════════╗
  ║  🎤 VOICE COMMANDS (or type them!)                                ║
  ╠═══════════════════════════════════════════════════════════════════╣
  ║  "Open Outlook"        - Opens Outlook                            ║
  ║  "Open Chrome"         - Opens Chrome                             ║
  ║  "Open VS Code"        - Opens Visual Studio Code                 ║
  ║  "Open Jira"           - Opens Bosch Jira                         ║
  ║  "Search [query]"      - Google search                            ║
  ║  "YouTube [query]"     - YouTube search                           ║
  ║  "Go to [website]"     - Open a website                           ║
  ║  "What time is it"     - Show current time                        ║
  ║  "Lock"                - Lock screen                              ║
  ║  "Exit" / "Quit"       - Close assistant                          ║
  ╚═══════════════════════════════════════════════════════════════════╝
    {Colors.END}""")

def execute_command(parsed, ps):
    action = parsed.get('action')
    
    if action == 'exit':
        response('Goodbye! 👋')
        return False
    
    elif action == 'help':
        show_help()
    
    elif action == 'open':
        print(f"{Colors.YELLOW}  ⚡ Opening...{Colors.END}")
        result = ps.open_app(parsed['app'])
        success(result)
    
    elif action == 'url':
        print(f"{Colors.YELLOW}  ⚡ Opening...{Colors.END}")
        result = ps.open_url(parsed['target'])
        success(result)
    
    elif action == 'search':
        print(f"{Colors.YELLOW}  ⚡ Searching...{Colors.END}")
        result = ps.search_web(parsed['query'], parsed.get('engine', 'google'))
        success(result)
    
    elif action == 'time':
        now = datetime.now()
        response(f"It's {now.strftime('%I:%M %p')} on {now.strftime('%B %d, %Y')}")
    
    elif action == 'lock':
        print(f"{Colors.YELLOW}  ⚡ Locking...{Colors.END}")
        ps.execute('rundll32.exe user32.dll,LockWorkStation')
        success('Locking screen...')
    
    elif action == 'command':
        print(f"{Colors.YELLOW}  ⚡ Executing...{Colors.END}")
        result = ps.execute(parsed['command'])
        print(f"{Colors.CYAN}\n  📤 Output:\n{Colors.WHITE}  {result}{Colors.END}")
    
    elif action == 'unknown':
        response(f'I don\'t understand "{parsed["input"]}". Say "help" for commands.')
    
    return True

def main():
    os.system('cls' if os.name == 'nt' else 'clear')
    banner()
    
    # Initialize
    ps = PowerShellTool()
    
    try:
        recognizer = WindowsSpeechRecognizer()
        voice_available = True
    except Exception as e:
        error(f"Voice not available: {e}")
        info("You can still type commands manually.")
        voice_available = False
    
    running = True
    
    while running:
        try:
            user_input = input(f"{Colors.GREEN}\n  [ENTER=🎤 Voice | Type command]: {Colors.END}").strip()
            
            # Empty input = voice mode
            if user_input == '' and voice_available:
                voice_text = recognizer.listen()
                if voice_text:
                    transcribed(voice_text)
                    parsed = parse_command(voice_text)
                    running = execute_command(parsed, ps)
            elif user_input:
                parsed = parse_command(user_input)
                running = execute_command(parsed, ps)
                
        except KeyboardInterrupt:
            response('Goodbye! 👋')
            break
        except Exception as e:
            error(str(e))

if __name__ == '__main__':
    main()
